src/thread_confined.cpp

```diff
@@ -0,0 +1,146 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "thread_confined.hpp"
+
+#include "impl/handover.hpp"
+
+using namespace realm;
+
+AnyThreadConfined::AnyThreadConfined(const AnyThreadConfined& thread_confined)
+{
+    switch (thread_confined.m_type) {
+        case Type::Object:
+            new (&m_object) Object(thread_confined.m_object);
+            break;
+
+        case Type::List:
+            new (&m_list) List(thread_confined.m_list);
+            break;
+
+        case Type::Results:
+            new (&m_results) Results(thread_confined.m_results);
+            break;
+    }
+    new (&m_type) Type(thread_confined.m_type);
+}
+
+AnyThreadConfined& AnyThreadConfined::operator=(const AnyThreadConfined& thread_confined)
+{
+    switch (thread_confined.m_type) {
```

Comment:
This also doesn't appear to destroy the existing state before initializing the new state.